erpnext/erpnext_integrations/doctype/leaflink_settings/leaflink_settings.py

```diff
@@ -0,0 +1,182 @@
+	if leaflink_settings.get("enable_leaflink") != 1:
+		frappe.throw("Please enable LeafLink in LeafLink Settings.")
+		return
+
+	#check if customer exists on leaflink
+	customers_url = leaflink_settings.get("customers_url")
+	customers_on_leaflink = requests.get(customers_url, headers=header, params={"name": customer_name}).json()
+
+	#create customer record on leaflink if it does not exist
+	if not customers_on_leaflink.get("count"):
+
+		#create new customer data
+		new_customer = {
+			"name": customer_name,
+			"owner": leaflink_settings.get("leaflink_id")
+	 	}
+
+		#post request to create customer in leaflink
+		post_response = requests.post(customers_url, headers=header, data=new_customer)
+		if post_response.status_code == 201:
+			frappe.msgprint("Customer successfully created on LeafLink!")
+		else:
+			frappe.msgprint("Could not create customer. Please try again.")
+	else:
+		frappe.msgprint("Customer already exists on LeafLink!")
+
+
+@frappe.whitelist(allow_guest=True)
+def sales_order_from_leaflink(**args):
+
+	#create integration request
+	integration_request = frappe.new_doc("Integration Request")
+	integration_request.update({
+		"integration_type": "Host",
+		"integration_request_service": "LeafLink",
+		"status": "Queued",
+		"data":  json.dumps(args, default=json_handler),
+		"reference_doctype": "Sales Order"
+	})
+	integration_request.save(ignore_permissions=True)
+
+	#curate data from settings for post request
+	leaflink_settings = frappe.get_cached_doc("LeafLink Settings")
+	header = {
+		leaflink_settings.get("api_key"): leaflink_settings.get("api_value")
+	}
+	sales_order = args.get("data")
+
+	#ensure leaflink is enabled before proceeding further.
+	if leaflink_settings.get("enable_leaflink") != 1:
+		integration_request.update({
+			"status": "Failed",
+			"error": "Leaflink is not enabled."
+		})
+		integration_request.save(ignore_permissions=True)
+		return
+
+	#validate incoming keys
+	key = bytes(leaflink_settings.get("incoming_api_authentication_key"), 'utf-8')
+	incoming_signature = frappe.local.request.headers['LL-Signature'].encode("utf-8")
+	request_body = frappe.local.request.data
+	computed_signature = base64.b64encode(hmac.new(key, request_body, digestmod=hashlib.sha256).digest())
+	success = (computed_signature == incoming_signature)
+	if not success:
+		integration_request.update({
+			"status": "Failed",
+			"error": "Signatures do not match. Please check the Incoming Key in LeafLink Settings."
+		})
+		integration_request.save(ignore_permissions=True)
+		return
+
+
+	#exit if the action is not to create a new sales order
+	if args.get("action") != "create":
+		return
+
+	#extract customer and license data
+	customer = sales_order.get("customer")
+	customer_name = customer.get('name')
+	license_no = customer.get('license_number')
+
+	#retrieve customer_name if license_no is in record
+	if not frappe.db.exists("Customer", customer_name):
+
+		license_no = customer.get('license_number')
```

Comment:
```suggestion
		license_no = customer.get('license_number')
```
Why defining variable when you can directly do object.get("field")

Reply:
Same as above